RFD 0189 Windows Desktop Access guide (#48013)
gravitational/teleport: rfd/0189-windows-access-guide.md

---
authors: Dave Sudia (david.sudia@goteleport.com)
state: draft
---

# RFD 0189 - Windows Desktop Access guide

## Required Approvers

* Engineering: @r0mant
* Product: @xinding33

## Overview

### Problem

* Currently only 15% of users convert from going to the docs to starting a session

> * We need better insight into drop-off points: we only know they clicked out to the docs then started a session and don’t know where they’re struggling

Reviewer: ```suggestion
* We need better insight into drop-off points: we only know they clicked out to the docs and whether they started a session, and don’t know where they’re struggling.
```

### Other data/context

> Historically we have more local user sessions than AD sessions (last 180 days, 36.3k local vs 30.1k AD). This may be skewed by us not getting metrics from major self-hosted customers who are likely using AD.

Reviewer: ```suggestion
Historically Teleport Cloud has more local user sessions than AD sessions (last 180 days, 36.3k local vs 30.1k AD). This may be skewed by us not getting metrics from major self-hosted customers who are likely using AD.
```

### Appetite/Resources

> We want the scope of this project to be doable for a full-stack engineer with some design input, within 6 weeks.

Reviewer: I don't think this is doable in 6 weeks.

We had a tighter-scoped goal of doing this only for EC2 instances on AWS a couple quarters ago and couldn't manage to fit that in with a dev who was experienced in the Windows domain.

This is more complicated than any of the existing discover flows that I'm aware of, because you need to install and configure Teleport software on at least two servers instead of 1 (the Linux instance running the agent and the target Windows instance).

Remotely installing software on a Windows machine is not something we've ever done, and this is further complicated by the facts that:
- the Windows host has to be rebooted after the software is installed
- the Teleport software that runs on Windows doesn't make any network connections, so it can't phone home and tell us when it's ready

None of this is impossible to solve, but it's tricky and likely requires a lot more than 6 weeks of a fullstack developer's time.

Author: @zmb3 I think I needed to be clearer about intention. I don't think we need to make this remotely installing the software. For other guides like SSH we just give them a script and have them run it. I think we should offer some kind of confirmation that the Desktop Service is running, but that can be a manual run of the script, and similarly running the script on the Windows box and restarting it can be manual actions on the user's part, we're just auto-generating the script.

Let me know what you think.

### Rabbit Holes

A thing that could potentially blow out the scope is trying to over-automate things for the user. To keep in scope as implementation progresses, we should favor adding more to the prerequisites that a user should set up over trying to write more complicated scripts.

We could try to do a guide for AD setup, but that is much more complex. There have been past attempts at making a wizard for AD setup that failed due to the complexity, and a decision was made that it should be referred to SE/IE teams.

### Out of Bounds

We don’t want to overhaul the Windows experience right now. It would be possible to streamline this process by not requiring a Windows Desktop Service running on a Linux box, and this is something that customers have asked for, but that would push us well out of our appetite for this project. We’ll revisit this later as another effort.

### Outlines/Sketches

An initial draft of the page flow is as follows:
* Before You Start
  * Steps overview
  * Prequisites
    * Windows machine accessible by a linux box
    * Must have hostname
* Set up Windows box
  * Generate script to install components (similar to linux ssh flow)

> * Setup Service

Reviewer: Something to consider -

This flow implies that we'll stand up a new agent (running on Linux) each time they enter the flow. This is probably fine for the very first getting started experience, but we often see customers struggle with what's next after that.

We don't want people to run through this flow a bunch of times, as having a Linux agent per-Windows host is an inefficient and expensive setup.

We can expect questions like "OK, I ran through that once and added a desktop. What do I do if I want to add a couple more, or if I want to enable discovery?"

Author: This is a good point. One thing I'd like to start adding to guides, but starting with this one, is links out to docs at the end for "dive deeper" or "next steps." Maybe part of creating this guide should also be creating a doc page that explains how to modify the Desktop Service to add more Windows hosts that we link to at the end?
